### PR TITLE
[BEAM-1422] ParDo should comply with PTransform style guide

### PR DESCRIPTION
**Pull request by jkff**
Backward-incompatible changes:
* Removes ParDo.Unbound and UnboundMulti (as a result, the only entry point is ParDo.of(DoFn) - you can no longer specify ParDo.withSideInputs(...).of(fn) and such)
* Renames ParDo.Bound to ParDo.SingleOutput and ParDo.BoundMulti to ParDo.MultiOutput

R: @kennknowles 
CC: @davorbonaci @dhalperi

### CODE REVIEW

**Comment by kennknowles** [comment]
This is unfortunately going to conflict heavily with #2145.

**Comment by jkff** [comment]
I'm aware of that PR and I think the conflicts won't be *that* bad. I'm fine submitting mine later and dealing with the conflicts - still worth a review for whether the proposed changes make sense.

**Comment by kennknowles** [comment]
From the high level description, it makes sense. I would consider inlining `ParDo.MultiOutput` to `ParDo` (no need for it to just be a namespace) since it is the true primitive but that's really a matter of style. I think the SDK tends not to do that.

**Comment by jkff** [comment]
Very strange, these tests pass locally for me.

**Comment by jkff** [comment]
retest this please

**Comment by kennknowles** [comment]
Unfortunately, the ParDo change that you've likely rebased on had to be reverted. If you can be patient, it will likely be rolled forwards identically but with an added override in the Dataflow runner to restore the prior behavior just for Dataflow.

**Comment by kennknowles** [comment]
I believe this has now been rolled forwards and you should proceed with the PR

**Comment by jkff** [comment]
Run Dataflow RunnableOnService

**Comment by jkff** [comment]
Run Dataflow RunnableOnService

**Comment by jkff** [comment]
Run Dataflow RunnableOnService

**Comment by jkff** [comment]
I sure hope this is flaky. Let's give it a shot:

**Comment by jkff** [comment]
Run Dataflow RunnableOnService

**Comment by jkff** [comment]
Run Dataflow RunnableOnService

**Comment by kennknowles** [comment]
The failing test was already broken, and recently fixed. I'm not sure if you caught the worker container image update that would fix it, but it doesn't block this PR either way.

**Comment by jkff** [comment]
Yeah, that's why I rebased - I expect it to succeed this time (and it succeeded when I ran it against a local instance of Dataflow). In that case, PTAL :)

**Comment by jkff** [comment]
R: @tgroh because Kenn says he has too many reviews in the backlog right now.

**Comment by kennknowles** [comment]
Also R: -@kennknowles

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectRunner.java [hunk 1]
```diff
@@ -168,7 +169,7 @@
   /** The set of {@link PTransform PTransforms} that execute a UDF. Useful for some enforcements. */
   private static final Set<Class<? extends PTransform>> CONTAINS_UDF =
       ImmutableSet.of(
-          Read.Bounded.class, Read.Unbounded.class, ParDo.Bound.class, ParDo.BoundMulti.class);
+          Read.Bounded.class, Read.Unbounded.class, ParDo.SingleOutput.class, MultiOutput.class);
```
Comment (jkff) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectRunner.java]: Ack, but it probably shouldn't be part of the current PR.

**Pull request closed without merging**